geo/src/algorithm/voronoi_diagram.rs

```diff
@@ -0,0 +1,970 @@
+use std::{fmt, result};
+
+use crate::{coord, Coord, GeoFloat, Line, MultiPoint, Point, Polygon, Rect, Triangle};
+
+use crate::triangulate_delaunay::{
+    DelaunayTriangle, DelaunayTriangulationError, DEFAULT_SUPER_TRIANGLE_EXPANSION,
+};
+use crate::{BoundingRect, TriangulateDelaunay};
+
+type Result<T> = result::Result<T, VoronoiDiagramError>;
+
+#[derive(Debug, Clone)]
+pub struct VoronoiComponents<T: GeoFloat> {
+    pub delaunay_triangles: Vec<Triangle<T>>,
+    pub vertices: Vec<Coord<T>>,
+    pub lines: Vec<Line<T>>,
+    pub neighbours: Vec<(Option<usize>, Option<usize>)>,
+}
+
+pub trait VoronoiDiagram<T: GeoFloat>
+where
+    f64: From<T>,
+{
+    fn compute_voronoi_components(
+        &self,
+        clipping_mask: Option<&Polygon<T>>,
+    ) -> Result<VoronoiComponents<T>>;
+    fn compute_voronoi_diagram(&self, clipping_mask: Option<&Polygon<T>>) -> Result<Vec<Polygon>>;
+}
+
+impl<T: GeoFloat> VoronoiDiagram<T> for Polygon<T>
+where
+    f64: From<T>,
+{
+    fn compute_voronoi_components(
+        &self,
+        clipping_mask: Option<&Polygon<T>>,
```

Comment:
`&Polygon<T>` should be `&Self`, here and in next fn def

Reply:
I've updated for the `Polygon<T>` implementation, but it can't be self for `MultiPoint<T>` as the clipping_mask is expected to be `Polygon<T>`.

Reply:
If the clipping mask should always be a `Polygon<T>`, even for other implementations, create a `type ClippingMask = Polygon<T>;` so it is clear that it is unrelated to the `Self`.